### ISSUE TITLE
ConvertValue function for godror.Number to bool

### ISSUE THREAD

**Issue by omerkirk** [godror/godror#39]
Hi Tamas,

Great job on the new driver.
database/sql package allows int to bool value conversion, however I think godror doesn't implement boolType conversions from godror.Number.  I am trying to upgrade a legacy code from goraclev1 to godror now, apparently goraclev1 had this feature so the code is riddled with int to bool conversions and I am getting the following error from many queries: 

`
sql/driver: couldn't convert 0 (godror.Number) into type bool 
`

Here is the database/sql ConvertValue implementation for boolType:

```
func (boolType) ConvertValue(src interface{}) (Value, error) {
	switch s := src.(type) {
	case bool:
		return s, nil
	case string:
		b, err := strconv.ParseBool(s)
		if err != nil {
			return nil, fmt.Errorf("sql/driver: couldn't convert %q into type bool", s)
		}
		return b, nil
	case []byte:
		b, err := strconv.ParseBool(string(s))
		if err != nil {
			return nil, fmt.Errorf("sql/driver: couldn't convert %q into type bool", s)
		}
		return b, nil
	}

	sv := reflect.ValueOf(src)
	switch sv.Kind() {
	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
		iv := sv.Int()
		if iv == 1 || iv == 0 {
			return iv == 1, nil
		}
		return nil, fmt.Errorf("sql/driver: couldn't convert %d into type bool", iv)
	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
		uv := sv.Uint()
		if uv == 1 || uv == 0 {
			return uv == 1, nil
		}
		return nil, fmt.Errorf("sql/driver: couldn't convert %d into type bool", uv)
	}

	return nil, fmt.Errorf("sql/driver: couldn't convert %v (%T) into type bool", src, src)
}
```

**Comment by tgulacsi** [comment]
Where should we put this ConvertValue method?
What if you Scan into such boolType ?

**Comment by omerkirk** [comment]
I am trying to figure out ColumnConverter, ValueConverter and NameValueChecker interfaces to be able to help you with a PR since I opened the issue but I couldn't understand how they come together. 

Since there is no place in the driver that knows the destination type, and ColumnConverter code in the database/sql calls driver.Bool.ConvertValue. I assumed that this is overrideable by each driver but it isn't because driver.Bool is not of type ValueConverter but some private type called boolType that is owned by database/sql. 

I didn't understand your second question, but my issue is the column type is Number in our db. But these values are either 1 or 0 since they are flag columns and I am trying to map these to bool properties in my structs. If I can't map 1s and 0s to bools either I will have to change the queries or change the destination property types to int which is what I am trying to avoid. 

I would like to help if you can point me in the right direction.

**Comment by omerkirk** [comment]
After some research I see that this has been discussed back in 2018 and it seems there is an accepted proposal but I couldn't find any development on it. 

https://github.com/golang/go/issues/22544

**Comment by omerkirk** [comment]
I think I found a dirty fix for this issue. I tried to use the fact that only the 0 and 1 values are valid for bool conversion, so I changed number case at Next function at rows.go 283 like the following

```
case C.DPI_ORACLE_TYPE_NUMBER:
			if isNull {
				//if Log != nil { Log("msg", "null", "i", i, "T", fmt.Sprintf("%T", dest[i]), "type", reflect.TypeOf(dest[i])) }
				dest[i] = nil
				continue
			}
			switch col.NativeType {
			case C.DPI_NATIVE_TYPE_INT64:
				dest[i] = int64(C.dpiData_getInt64(d))
			case C.DPI_NATIVE_TYPE_UINT64:
				dest[i] = uint64(C.dpiData_getUint64(d))
			case C.DPI_NATIVE_TYPE_FLOAT:
				//dest[i] = float32(C.dpiData_getFloat(d))
				dest[i] = printFloat(float64(C.dpiData_getFloat(d)))
			case C.DPI_NATIVE_TYPE_DOUBLE:
				//dest[i] = float64(C.dpiData_getDouble(d))
				dest[i] = printFloat(float64(C.dpiData_getDouble(d)))
			default:
				b := C.dpiData_getBytes(d)
				s := C.GoStringN(b.ptr, C.int(b.length))
				if r.NumberAsString() {
					dest[i] = s
				} else {
					if s == "0" || s == "1" {
						num, _ := strconv.ParseInt(s, 10, 8)
						dest[i] = num
					} else {
						dest[i] = Number(s)
					}
				}
				if Log != nil {
					Log("msg", "b", "i", i, "ptr", b.ptr, "length", b.length, "typ", col.NativeType, "int64", C.dpiData_getInt64(d), "dest", dest[i])
				}
			}
			if Log != nil {
				Log("msg", "num", "t", col.NativeType, "i", i, "dest", fmt.Sprintf("%T %+v", dest[i], dest[i]))
			}
```

This part is new:

```
if s == "0" || s == "1" {
    num, _ := strconv.ParseInt(s, 10, 8)
    dest[i] = num
} else {
    dest[i] = Number(s)
}
```

With this change if the Number value is eligible for bool conversion than the destination type is not godror.Number but instead int8 but could be any type that the database/sql knows how to convert. 

It is a dirty fix and I didn't test any possible side effects yet.

**Comment by tgulacsi** [comment]
That's a very dirty fix that fails everywhere else (why does it return int8(1) but Number(2) ?).

What if you create a BoolType with proper ConvertValue method and Scan into it?

**Comment by tgulacsi** [comment]
See https://pkg.go.dev/github.com/godror/godror/?tab=doc#BoolToString for example.

**Comment by omerkirk** [comment]
Yes İ agree it is not a good solution, what if the code detected if the number is an integer and returns int64 if it is. 

A custom type and handle scanning from goror.Number is a cleaner way,  I am just trying to understand if this can be handled in the driver rather then the user land.

**Comment by tgulacsi** [comment]
But you only need a
```
type Booler bool
var _ sql.Scanner = Booler{}
func (b Booler) Scan(src interface{}) error {
  switch src := src.(type) {
    case int: *b = x == 1
    case string: *b = x == "Y" || x == "T"  // or any string your database model treats as truth value
    default: return fmt.Errorf("unknown scanner source %T", src)
  }
  return nil
}
```
and
```
rows, err := db.QueryContext(...)
var mystruct struct {
  b bool
}
for rows.Next() {
  rows.Scan((*Booler)(&mystruct.b))
}
```
Why is this too much?

**Comment by omerkirk** [comment]
I use sqlx for the mapping, so I don't have control over the scanning part. This currently doesn't work without changing the struct properties from bool to Booler, which then changes how they are handled in the code. 

It would cleanup the user code a lot if this can be handled in the driver, which database/sql actually supports (int -> bool conversion). The only problem is the sql package doesn't let the driver author to provide a way to convert from custom types into driver.Value types. If we can somehow minimize the cases where the driver has to return a custom type then we can take advantage of the database/sql support in more cases. 

The case I changed in the above comment is the case where the driver cannot infer the type from the DB column type, but it still can from the value and return database/sql known type without any loss of information.  I am still not saying this is the ideal solution because it might have side effects, but I still think this case can be handled a lot cleaner in the driver than in the user code.

**Comment by tgulacsi** [comment]
Then maybe an Option that changes the behaviour for INT types?

I won't change the default behaviour, and I hate the  suggested "the returned type is sometimes int, sometimes godror.Number" hack.

As database/sql.convertAssignRows uses driver.Bool.ConvertValue, another solution arises:
use godror.NumberAsString() option to force every number to be a string - that can be converted to bool by driver.boolType.ConvertValue.

So no change needed in the driver, just use godror.NumberAsString().

**Comment by omerkirk** [comment]
Yes this works, apparently strconv.ParseBool can convert from "1" and "0". Is there any disadvantages for using string instead of godror.Number ? If there isn't is there any reason why returning string isn't the default behaviour and if the godror.Number type is necessary in any way. 

Thanks a lot for your help, I am closing the issue.

**Comment by tgulacsi** [comment]
PTAL the "nonumber" branch - I've no objections for simplifying the code and return number as strings wherever possible.

We can't get rid of godror.Number as we need to distinguins destination targets, though.

**Comment by tgulacsi** [comment]
It's merged to master and tagged as v0.14.0.
